### ISSUE TITLE
Marginally better LUB for different instantiations of the same interface.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#4204]
Currently **Up**(T1, T2)` recognizes if `T1` is `C<S1,..,Sn>` and `T2` is `C<R1,...,Rn>`, with some `Si` != `Ri`,  and tries to unify the type parameters.
However if `T1` *implements* `C<S1,...,Sn>` and `T2` *implements* `C<R1,...,Rn>`, then we may fall through to the old depth-based LUB algorithm, which only looks at *common types* between the two types, and `C<S1,...,Sn>` is a different type than `C<R1,...,Rn>`, so neither gets included in the comparison.

I suggest that we reconsider that approach. When looking for common supertypes, we'll find `C1<S1,...,Sn>` and `C1<R1,...,Rn>` has the same *depth*, but are not equal. We should then realize that they are *related*, and **if one is a supertype of the other** (and not vice versa), then we *do* include the supretype in the set of common supertypes.

Concretely I'm trying to add a type parameter to `Pattern`, so that `Pattern<M extends Match>` has methods returning `M`, and then `RegExp implements Pattern<RegExpMatch>`.
My problem is that `["a", RegExp("b")]` used to be inferred as `List<Pattern>`, but with the type parameter, it is inferred as `List<Object>` because `Pattern<Match>` and `Pattern<RegExpMatch>` are considered completely unrelated.

(We can probably do much better if we built a LUB from scratch, but this particular change seems within reach and still able to do something good in some cases.)

**Comment by mateusfccp** [comment]
Is this related to this case?

```dart
void main() {
  int value = 0;

  // positiveInt is inferred as `Object`
  final positiveInt = switch (value) {
    > 0 => Some(value),
    _ => None(),
  };

  // positiveInt is inferred as `Option<int>`, as expected
  final positiveInt = switch (value) {
    > 0 => Some(value),
    _ => none,
  };
}

sealed class Option<T> {
  const Option._();
}

final class Some<T> extends Option<T> {
  const Some(this.value) : super._();

  final T value;
}

final class None<T> extends Option<T> {
  const None() : super._();
}

const Option<Never> none = None();
```

**Comment by lrhn** [comment]
Yes, precisely. The context of the `positiveInt` switch is empty, so each branch is inferred by itself, to `Some<int>` and `None<dynamic>`, which implement `Option<int>` and `Option<dynamic>`, which are not the same type, and therefore there is no shared supertype other than `Object`.
(Even if you made `None` be `final class None extends Option<Never>`, you'd just have `Option<Never>` which is still not `Option<int>`.)

The proposal here is to be just a bit less obtuse when doing intersection of types.
If one set contains `Foo<X1, ..., Xn>` and the other set contains `Foo<Y1,...,Yn>`, then recognize this and try something *simple* (and not a recursive LUB that could diverge): For each Xi, Yi, if one is a supertype of the other, retain that. (With only covariant variables, I think that the result will be well-bounded if the original types are.)

Can even get ambitious and define a "point-wise max", which *potentially* finds a type for two different types by only structural recursion and supertype checks: **MAX**(`S`, `T`) is:
* If `S` \<: `T` and not `S` \<: `T` then result is `T`.
* If `T` \<: `S` and not `S` \<: `T` then result is `S`.
* If `T` \<: `S` and `S` \<: `T` then
    * if **TOP**(`T`) then if **MORETOP**(`T`, `S`) then `T` else `S`.
    * if **BOTTOM**(`T`) then `Never`
    * Else the result is undefined.
* If `S` is `S0?` and `T` is `T0?` and `R` = **MAX**(`S0`, `T`) is defined, then `R` if `R` is `nullable` and `R?` if not.
* If `S` is `S0?` and `T` is not `T0?` for any `T0`, and `R` = **MAX**(`S0`, `T`) is defined, then `R` if `R` is `nullable` and `R?` if not.
* If `S` is not `S0?` for any `S0` and `T` is `T0?`, and `R` = **MAX**(`S0`, `T`) is defined, then `R` if `R` is `nullable` and `R?` if not.
* If `S` is `C<S1,...,Sn>` and `T` is `C<T1,...,Tn>` then let `Ri` = **MAX**(`Si`, `Ti`) for each `i` in 1..n. If all of these exist,
   the result is `C<R1,...,Rn>`.
* If `S` is `C<S1, ..., Sn>` and `T` implements `C<T1, ..., Tn>`, same as above.
* If `T` is `C<T1, ..., Tn>` and `S` implements `C<S1, ..., Sn>`, same as above.
* If `S` and `T` are records with the same shape, result is record with that shape where
   each field is the **MAX** of the two field types.
* If `S` and `T` are functions with the same or compatible shapes (handwave: if one has an optional parameter and the other
   has a required or no parameter, then result parameter is required or absent. If both have parameter, type is **MIN**
   of their types, which is defined parallel to **MAX**), return type is **MAX** of return types, and type parameters ...
   probably has to be the same. In every case, if a **MAX** or **MIN** is undefined so is the result.
* `S` is `S0?` and `T` is `T0?`, then result is `R?` where `R` is **MAX**(`S0`,`T0`) if that is defined.
* ... etc ... (aka. this is where it starts getting hard, so I'll have to use more time to be precise.)

The point is to *only* be structural in the type terms, which avoids the infinite recursion that can be caused by finding super-interfaces. For each structure, try to find a maximal type by combining only the component types that occur in the of the original types, no new types invented. If at any level, one is a subtype of the other, stop there and use it.

**Comment by mateusfccp** [comment]
> (Even if you made None be final class None extends Option<Never>, you'd just have Option<Never> which is still not Option<int>.)

Shouldn't `Option<Never>` be a subtype of `Option<int>`? Then it would be considered in the LUB algorithm, as both are `Option`, `LUB(Never, int)` which would be `int`, leading to `Option<int>`? I don't get why this shouldn't work.

**Comment by lrhn** [comment]
The problem with the LUB algorithm is that it only starts looking at superinterfaces at the end, and when it does that, it only checks for the *same* superinterface, not one being a subtype of the other.

The inputs to **UP** are `Some<int>` and `None<dynamic>` (or `None<Never>`, it makes no difference as long as it's not `None<int>`).

The **UP** starts by structurally looking at the types, but they two types are not structurally related.
Then if falls through to the old **LUB** algorithm which takes *all the superinterfaces* of the two types, and creates an intersection to find the superinterfaces of both. The superinterfaces of `Some<int>` are `Option<int>` and `Object`. The superinterfaces of `None<dynamic>` are `Option<dynamic>` and `Object`. The only type in both sets is `Object`, so that's the upper bound.

It's that last step where I'd want to do something just a little less oblivious.

> Did you mean ...

I did. fixed.

**Comment by mateusfccp** [comment]
@lrhn Is the LUB algorithm specified anywhere or I would have to check the CFE code? I've seen that the concept of LUB is specified in the Dart's language specification, but it seems the algorithm doesn't follow it strictly?

**Comment by lrhn** [comment]
The **UP** algorithm is here: https://github.com/dart-lang/language/blob/main/resources/type-system/upper-lower-bounds.md
I hope it's up-to-date, but @stereotype441 gives it an occasional improvement.

At the end, it falls back on the "LUB" algorithm that has been almost the same since Dart 1.
It was last modified with extension types: https://github.com/dart-lang/language/blob/main/accepted/future-releases/extension-types/feature-specification.md#changes-to-other-types-and-subtyping

Other than that I think the definition in the language specification is correct.
